patroni/postgresql/slots.py

```diff
@@ -47,65 +95,178 @@ def drop_replication_slot(self, name):
-                                self._schedule_load_slots = True
                 self._replication_slots = slots
             except Exception:
                 logger.exception('Exception when changing replication slots')
                 self._schedule_load_slots = True
+        return ret
+
+    @contextmanager
+    def _get_leader_connection_cursor(self, leader):
+        conn_kwargs = leader.conn_kwargs(self._postgresql.config.rewind_credentials)
+        conn_kwargs['database'] = self._postgresql.database
+        with get_connection_cursor(connect_timeout=3, options="-c statement_timeout=2000", **conn_kwargs) as cur:
+            yield cur
+
+    def check_logical_slots_readiness(self, cluster, nofailover, replicatefrom):
+        if self._unready_logical_slots:
+            slot_name = cluster.get_my_slot_name_on_primary(self._postgresql.name, replicatefrom)
+            try:
+                with self._get_leader_connection_cursor(cluster.leader) as cur:
+                    cur.execute("SELECT catalog_xmin FROM pg_catalog.pg_get_replication_slots()"
+                                " WHERE NOT pg_catalog.pg_is_in_recovery() AND slot_name = %s", (slot_name,))
+                    if cur.rowcount < 1:
+                        return logger.warning('Physical slot %s does not exist on the primary', slot_name)
+                    catalog_xmin = cur.fetchone()[0]
+            except Exception as e:
+                return logger.error("Failed to check %s physical slot on the primary: %r", slot_name, e)
+        for name in list(self._unready_logical_slots):
+            value = self._replication_slots.get(name)
+            if not value or catalog_xmin <= value['catalog_xmin']:
+                self._unready_logical_slots.remove(name)
+                if value:
+                    logger.info('Logical slot %s is safe to be used after a failover', name)
+
+    def copy_logical_slots(self, leader, slots):
+        with self._get_leader_connection_cursor(leader) as cur:
+            try:
+                cur.execute("SELECT slot_name, catalog_xmin, "
+                            "pg_catalog.pg_wal_lsn_diff(confirmed_flush_lsn, '0/0')::bigint, "
+                            "pg_catalog.pg_read_binary_file('pg_replslot/' || slot_name || '/state')"
+                            " FROM pg_catalog.pg_get_replication_slots() WHERE NOT pg_catalog.pg_is_in_recovery()"
+                            " AND slot_name = ANY(%s)", (slots,))
+                slots = {r[0]: {'catalog_xmin': r[1], 'confirmed_flush_lsn': r[2], 'data': r[3]} for r in cur}
+            except Exception as e:
+                logger.error("Failed to copy logical slots from the %s via postgresql connection: %r", leader.name, e)
+
+        if isinstance(slots, dict) and self._postgresql.stop():
+            for name, value in slots.items():
+                slot_dir = os.path.join(self._postgresql.data_dir, 'pg_replslot', name)
+                os.makedirs(slot_dir)
```

Comment:
Is it possible to do this in more "atomical" manner? If I understand correctly, if patroni will for whatever reason die after creating such directory, Postgres will panic and patroni will not be able to start it without cleaning up first.

Reply:
Addressed in https://github.com/zalando/patroni/pull/1820/commits/88070569ef2a184ae84b6a3762bad56a949c9a42